api/src/certification/complementary-certification/domain/usecases/attach-badges.js

```diff
@@ -0,0 +1,153 @@
+    (badgeToAttachDTO) => {
+      const badgeToAttach = BadgeToAttach.from({
+        ...badgeToAttachDTO,
+        complementaryCertificationId,
+        userId,
+      });
+      return badgeToAttach;
+    },
+  );
+
+  return DomainTransaction.execute(async (domainTransaction) => {
+    const relatedComplementaryCertificationBadgesIds =
+      await complementaryCertificationBadgesRepository.getAllIdsByTargetProfileId({
+        targetProfileId: targetProfileIdToDetach,
+      });
+
+    await _detachExistingComplementaryCertificationBadge({
+      complementaryCertificationBadgesRepository,
+      relatedComplementaryCertificationBadgesIds,
+      domainTransaction,
+    });
+
+    for (const badgeToAttach of complementaryCertificationBadgesToAttach) {
+      const complementaryCertificationBadge = BadgeToAttach.from({
+        ...badgeToAttach,
+        complementaryCertificationId,
+        userId,
+      });
+      await _attachNewComplementaryCertificationBadge({
+        complementaryCertificationBadgesRepository,
+        complementaryCertificationBadge,
+        domainTransaction,
+      });
+    }
+  });
+};
+
+export { attachBadges };
+
+async function _attachNewComplementaryCertificationBadge({
+  complementaryCertificationBadgesRepository,
+  complementaryCertificationBadge,
+  domainTransaction,
+}) {
+  await complementaryCertificationBadgesRepository.attach({
+    complementaryCertificationBadge,
+    domainTransaction,
+  });
+}
+
+async function _detachExistingComplementaryCertificationBadge({
+  complementaryCertificationBadgesRepository,
+  relatedComplementaryCertificationBadgesIds,
+  domainTransaction,
+}) {
+  if (relatedComplementaryCertificationBadgesIds.length === 0) {
+    throw new NotFoundError('No badges for this target profile.');
+  }
+
+  await complementaryCertificationBadgesRepository.detachByIds({
+    complementaryCertificationBadgeIds: relatedComplementaryCertificationBadgesIds,
+    domainTransaction,
+  });
+}
+
+function _compareLevels(level1, level2) {
+  return level1 - level2;
+}
+
+function _ifEveryBadgeHasALevelThatIsUniq({ sortedUniqLevels, complementaryCertificationBadgesToAttachDTO }) {
+  return sortedUniqLevels.length == complementaryCertificationBadgesToAttachDTO.length;
+}
+
+function _isFirstLevelDifferentThanOne(sortedUniqLevels) {
+  return sortedUniqLevels[0] !== 1;
+}
+
+function _isLastLevelDifferentThanExpectedMaximum({ sortedUniqLevels, complementaryCertificationBadgesToAttachDTO }) {
+  return sortedUniqLevels[sortedUniqLevels.length - 1] !== complementaryCertificationBadgesToAttachDTO.length;
+}
+
+function _verifyThatLevelsAreConsistent({ complementaryCertificationBadgesToAttachDTO }) {
+  const extractedLevelsFromBadges =
+    complementaryCertificationBadgesToAttachDTO?.map((badge) => badge.level).filter(Number.isInteger) ?? [];
+  const sortedUniqLevels = uniq([...extractedLevelsFromBadges]).sort(_compareLevels);
```

Comment:
est-ce qu'un simple sort n'aurait pas suffit?

uniq([...extractedLevelsFromBadges]).sort();

Reply:
KO si > 10

> 
> Sort Compare Function
> Sorting alphabetically works well for strings ("Apple" comes before "Banana").
> But, sorting numbers can produce incorrect results.
> "25" is bigger than "100", because "2" is bigger than "1".
> You can fix this by providing a "compare function" (See examples below).

Reply:
Par contre on pourrait faire 
```suggestion
  const sortedUniqLevels = uniq([...extractedLevelsFromBadges]).sort((level1, level2) => level1 - level2;);
```